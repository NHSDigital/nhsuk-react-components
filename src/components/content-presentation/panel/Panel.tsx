import classNames from 'classnames';
import { Children, forwardRef, type ComponentPropsWithoutRef, type FC } from 'react';
import { HeadingLevel, type HeadingLevelProps } from '#components/utils/HeadingLevel.js';
import { childIsOfComponentType } from '#util/types/TypeGuards.js';

export type PanelTitleProps = HeadingLevelProps;

const PanelTitle: FC<PanelTitleProps> = ({ children, headingLevel = 'h1', ...rest }) => (
  <HeadingLevel className="nhsuk-panel__title" headingLevel={headingLevel} {...rest}>
    {children}
  </HeadingLevel>
);

export type PanelProps = ComponentPropsWithoutRef<'div'>;

const PanelComponent = forwardRef<HTMLDivElement, PanelProps>(
  ({ children, className, ...rest }, forwardedRef) => {
    const items = Children.toArray(children);
    const title = items.find((child) => childIsOfComponentType(child, PanelTitle));
    const bodyItems = items.filter((child) => !childIsOfComponentType(child, PanelTitle));

    return (
      <div className={classNames('nhsuk-panel', className)} ref={forwardedRef} {...rest}>
        {title}
        {bodyItems ? <div className="nhsuk-panel__body">{bodyItems}</div> : null}
      </div>
    );
  },
);

PanelComponent.displayName = 'Panel';
PanelComponent.displayName = 'Panel.Title';

export const Panel = Object.assign(PanelComponent, {
  Title: PanelTitle,
});
