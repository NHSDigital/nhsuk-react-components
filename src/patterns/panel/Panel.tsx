import React, { FC, HTMLProps } from 'react';
import classNames from 'classnames';
import { AsElementLink } from '@util/types/LinkTypes';
import Card from '@components/navigation/card';
import { CardHeadingProps } from '@components/navigation/card/components/CardHeading';

interface PanelProps extends HTMLProps<HTMLDivElement> {
  labelProps?: HTMLProps<HTMLHeadingElement> & CardHeadingProps;
  backToTop?: boolean;
  backToTopButtonText?: string;
  backToTopLink?: string;
}

interface Panel extends FC<PanelProps> {
  LinkItem: FC<HTMLProps<HTMLAnchorElement>>;
  Item: FC<HTMLProps<HTMLLIElement>>;
}

const Panel: Panel = ({
  children,
  label,
  labelProps,
  backToTop,
  backToTopLink,
  backToTopButtonText,
}) => (
  <>
    <Card cardType="feature">
      <Card.Content>
        {label ? <Card.Heading {...labelProps}>{label}</Card.Heading> : null}
        <ul className="nhsuk-list nhsuk-list--border">{children}</ul>
      </Card.Content>
    </Card>

    {backToTop ? (
      <div className="nhsuk-back-to-top">
        <a className="nhsuk-back-to-top__link" href={backToTopLink}>
          {backToTopButtonText || 'Back to top'}
        </a>
      </div>
    ) : null}
  </>
);

const PanelItem: FC<HTMLProps<HTMLLIElement>> = ({ className, ...rest }) => (
  <li className={classNames('nhsuk-list-panel__item', className)} {...rest} />
);

const PanelLinkItem: FC<AsElementLink<HTMLAnchorElement>> = ({
  className,
  asElement: Component = 'a',
  ...rest
}) => (
  <PanelItem>
    <Component className={classNames('nhsuk-list-panel__link', className)} {...rest} />
  </PanelItem>
);

Panel.LinkItem = PanelLinkItem;
Panel.Item = PanelItem;

export default Panel;
