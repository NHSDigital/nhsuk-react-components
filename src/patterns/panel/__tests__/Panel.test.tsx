import React from 'react';
import { render } from '@testing-library/react';
import Panel from '..';

describe('Panel', () => {
  it('matches snapshot', () => {
    const { container } = render(<Panel />);

    expect(container).toMatchSnapshot('ListPanel');
  });

  it('renders label', () => {
    const { container } = render(<Panel label="Label" labelProps={{ className: 'test-label' }} />);
    const label = container.querySelector('.test-label');

    expect(label).toBeTruthy();
    expect(label?.textContent).toBe('Label');
  });

  it('renders back to top button', () => {
    const { container } = render(<Panel backToTop />);

    expect(container.querySelector('.nhsuk-back-to-top')).toBeTruthy();
  });

  it('does not render back to top button', () => {
    const { container } = render(<Panel />);

    expect(container.querySelector('.nhsuk-back-to-top')).toBeNull();
  });

  it('renders back to top button with custom text', () => {
    const { container } = render(<Panel backToTop backToTopButtonText="Custom" />);

    expect(container.querySelector('.nhsuk-back-to-top__link')?.textContent).toBe('Custom');
  });

  describe('ListPanel.Item', () => {
    it('matches snapshot', () => {
      const { container } = render(<Panel.Item />);

      expect(container).toMatchSnapshot('ListPanel.Item');
    });
  });

  describe('ListPanel.LinkItem', () => {
    it('matches snapshot', () => {
      const { container } = render(<Panel.LinkItem />);

      expect(container).toMatchSnapshot('ListPanel.LinkItem');
    });
  });
});
