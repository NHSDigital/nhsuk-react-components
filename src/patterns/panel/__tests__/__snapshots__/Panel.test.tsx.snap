// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel ListPanel.Item matches snapshot: ListPanel.Item 1`] = `
<div>
  <li
    class="nhsuk-list-panel__item"
  />
</div>
`;

exports[`Panel ListPanel.LinkItem matches snapshot: ListPanel.LinkItem 1`] = `
<div>
  <li
    class="nhsuk-list-panel__item"
  >
    <a
      class="nhsuk-list-panel__link"
    />
  </li>
</div>
`;

exports[`Panel matches snapshot: ListPanel 1`] = `
<div>
  <div
    class="nhsuk-card nhsuk-card--feature"
  >
    <div
      class="nhsuk-card__content nhsuk-card__content--feature"
    >
      <ul
        class="nhsuk-list nhsuk-list--border"
      />
    </div>
  </div>
</div>
`;
